lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -4426,23 +4710,500 @@ class JVPEmitter final
+          auto *lastArg = diffBB->getArguments().back();
+          assert(lastArg->getType() == diffStructLoweredType);
+          differentialStructArguments[&origBB] = lastArg;
+        }
+      }
+
+      LLVM_DEBUG({
+        auto &s = getADDebugStream()
+                  << "Original bb" + std::to_string(origBB.getDebugID())
+                  << ": To differentiate or not to differentiate?\n";
+        for (auto &inst : origBB) {
+          s << (differentialInfo.shouldDifferentiateInstruction(&inst)
+                    ? "[∂] " : "[ ] ")
+            << inst;
+        }
+      });
+    }
+
+    assert(diffBBMap.size() == 1 &&
+           "Can only currently handle single basic block functions");
+
+    // The differential function has type:
+    // (arg0', ..., argn', entry_df_struct) -> result'.
+    auto diffParamArgs =
+        differential.getArgumentsWithoutIndirectResults().drop_back();
+    assert(diffParamArgs.size() ==
+           attr->getIndices().parameters->getNumIndices());
+    auto origParamArgs = original->getArgumentsWithoutIndirectResults();
+
+    // Check if result is not varied.
+    SmallVector<SILValue, 8> origFormalResults;
+    collectAllFormalResultsInTypeOrder(*original, origFormalResults);
+    auto origResult = origFormalResults[getIndices().source];
+    // Emit warning if original result is not varied, because it will always
+    // have a zero derivative.
+    if (!activityInfo.isVaried(origResult, getIndices().parameters)) {
+      // Emit fixit if original result has a valid source location.
+      auto startLoc = origResult.getLoc().getStartSourceLoc();
+      auto endLoc = origResult.getLoc().getEndSourceLoc();
+      if (startLoc.isValid() && endLoc.isValid()) {
+        context.diagnose(startLoc, diag::autodiff_nonvaried_result_fixit)
+            .fixItInsert(startLoc, "withoutDerivative(at:")
+            .fixItInsertAfter(endLoc, ")");
+      }
+    }
+
+    // Create a mapping of the parameters.
+    auto autoDiffIndex = getIndices().parameters->begin();
+    for (auto index : range(diffParamArgs.size())) {
+      auto *diffParam = diffParamArgs[index];
+      auto *origParam = origParamArgs[*autoDiffIndex];
+      autoDiffIndex++;
+      if (diffParam->getType().isAddress()) {
+        setTangentBuffer(origEntry, origParam, diffParam);
+      } else {
+        setTangentValue(
+            origEntry, origParam, makeConcreteTangentValue(diffParam));
+      }
+      LLVM_DEBUG(getADDebugStream()
+                 << "Assigned parameter " << *diffParam
+                 << " as the tangent of original result " << *origParam);
+    }
+
+    // If there are indirect results, create a mapping.
+    auto origIndResults = original->getIndirectResults();
+    auto diffIndResults = differential.getIndirectResults();
+    assert(origIndResults.size() == diffIndResults.size());
+
+    for (auto &origBB : *original) {
+      for (auto i : range(diffIndResults.size())) {
```

Comment:
```suggestion
      for (auto i : indices(diffIndResults)) {
```

or if not compiling,

```suggestion
      for (auto i : swift::indices(diffIndResults)) {
```